encryption.go

```diff
@@ -0,0 +1,413 @@
+and correct headers
+
+:param endpoint: The endpoint/server to prepare requests to
+:param message: The unencrypted message to send to the server
+:return: A prepared request that has an decrypted message
+*/
+func (e *Encryption) PrepareEncryptedRequest(client *Client, endpoint string, message []byte) (string, error) {
+	url, err := url.Parse(endpoint)
+	if err != nil {
+		return "", err
+	}
+	host := strings.Split(url.Hostname(), ":")[0]
+
+	var content_type string
+	var encrypted_message []byte
+
+	if e.protocol == "credssp" && len(message) > e.SIXTEN_KB {
+		content_type = "multipart/x-multi-encrypted"
+		encrypted_message = []byte{}
+		message_chunks := [][]byte{}
+		for i := 0; i < len(message); i += e.SIXTEN_KB {
+			message_chunks = append(message_chunks, message[i:i+e.SIXTEN_KB])
+		}
+		for _, message_chunk := range message_chunks {
+			encrypted_chunk := e.encryptMessage(message_chunk, host)
+			encrypted_message = append(encrypted_message, encrypted_chunk...)
+		}
+	} else {
+		content_type = "multipart/encrypted"
+		encrypted_message = e.encryptMessage(message, host)
+	}
+
+	encrypted_message = append(encrypted_message, e.MIME_BOUNDARY...)
+	encrypted_message = append(encrypted_message, []byte("--\r\n")...)
+
+	req, err := http.NewRequest("POST", endpoint, bytes.NewBuffer(encrypted_message))
+	if err != nil {
+		return "", err
+	}
+
+	req.Header.Set("User-Agent", "Bob WinRM client")
+	req.Header.Set("Connection", "Keep-Alive")
+	req.Header.Set("Content-Length", fmt.Sprintf("%d", len(encrypted_message)))
+	req.Header.Set("Content-Type", content_type+";protocol=\""+string(e.protocolString)+"\";boundary=\"Encrypted Boundary\"")
+
+	resp, err := e.ntlmhttp.Do(req)
+	if err != nil {
+		return "", fmt.Errorf("unknown error %w", err)
+	}
+
+	body, err := e.ParseEncryptedResponse(resp)
+
+	return string(body), err
+}
+
+/*
+Takes in the encrypted response from the server and decrypts it
+
+:param response: The response that needs to be decrytped
+:return: The unencrypted message from the server
+*/
+func (e *Encryption) ParseEncryptedResponse(response *http.Response) ([]byte, error) {
+	contentType := response.Header.Get("Content-Type")
+	if strings.Contains(contentType, fmt.Sprintf(`protocol="%s"`, e.protocolString)) {
+		return e.decryptResponse(response, response.Request.URL.Hostname())
+	}
+	body, err := ioutil.ReadAll(response.Body)
+	response.Body.Close()
+	if err != nil {
+		return nil, err
+	}
+	return body, nil
+}
+
+func (e *Encryption) encryptMessage(message []byte, host string) []byte {
+	messageLength := []byte(fmt.Sprintf("%d", len(message)))
+	encryptedStream, _ := e.buildMessage(message, host)
+
+	messagePayload := bytes.Join([][]byte{
+		e.MIME_BOUNDARY,
+		[]byte("\r\n"),
+		[]byte("\tContent-Type: " + string(e.protocolString) + "\r\n"),
+		[]byte("\tOriginalContent: type=application/soap+xml;charset=UTF-8;Length=" + string(messageLength) + "\r\n"),
+		e.MIME_BOUNDARY,
+		[]byte("\r\n"),

```

Comment:
we could have a constant covering `MIME_BOUNDARY+"\r\n"` since it's used a lot of time.
Also one for `\r\n` :)

Reply:
I have the feeling the whole message payload could be done through a `Sprintf` and then transformed to bytes. It would be more readable.

Reply:
refactored